File: Andriod/Constructor/construction/src/main/java/com/palitri/openiot/construction/framework/tools/threading/ContinuousThread.java
package com.palitri.openiot.construction.framework.tools.threading;

public class ContinuousThread extends Thread {
    private volatile boolean running;
    private volatile boolean done;

    public volatile int interval;

    public ContinuousThread() {
        this.interval = 50;
    }

    @Override
    public void run() {
        this.running = true;
        this.done = false;

        try {
            while (this.running && this.action()) {
                Thread.sleep(this.interval);
            }
        }
        catch (InterruptedException ex) {
            return;
        }

        this.done = true;
    }

    public void terminate(boolean wait) {
        this.running = false;

        while (wait && !this.done)
            try {
                Thread.sleep(this.interval);
            }
            catch (InterruptedException ex) {
                return;
            }

    }

    public boolean action()
    {
        return false;
    }
}

File: Windows/OpenIoT.App/OpenIoT.Lib/Tools/Threading/ContinuousThread.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace OpenIoT.Lib.Tools.Threading
{
    public delegate bool ContinuousThreadAction();

    internal class ContinuousThread
    {
        private volatile bool loop;
        private volatile bool running;

        public volatile int interval;

        private readonly Thread thread;
        private readonly ContinuousThreadAction? action;

        public bool IsRunning => this.running;
        public bool IsClosing => this.running && !this.loop;

        public ContinuousThread(ContinuousThreadAction? action = null, int interval = 50)
        {
            this.thread = new Thread(this.Run);

            this.action = action;
            this.interval = interval;

            this.loop = false;
            this.running = false;
        }

        public void Start(bool continuous = true)
        {
            this.loop = continuous;

            this.thread.Start();
        }
        
        private void Run()
        {
            this.running = true;

            try
            {
                while (this.Action())
                {
                    if (!this.loop)
                        break;

                    Thread.Sleep(this.interval);
                }
            }
            catch
            {
            }
            finally
            {
                this.running = false;
            }
        }

        public void Terminate(bool wait)
        {
            this.loop = false;

            if (wait)
                this.Wait();
        }

        public void Wait()
        {
            while (this.running)
                try
                {
                    Thread.Sleep(this.interval);
                }
                catch (Exception)
                {
                }
        }

        public bool Action()
        {
            if (this.action == null)
                return false;

            return this.action.Invoke();
        }
    }

}
